File: examples/iotApp/ContentDirectory.cpp
//******************************************************************
//
// Copyright 2016 Intel Corporation All Rights Reserved.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

#include "ContentDirectory.h"
#include "IotivityUtility.h"

#include <chrono>

using namespace std;
using namespace OC;

ContentDirectory::ContentDirectory() :
    m_searchCaps{""},
    m_sortCaps{""},
    m_id{0},
    m_resetToken{""},
    m_featureList{""},
    m_resource{nullptr}
{
    m_browseResult = {"", 0, 0, 0};
    m_searchResult = {"", 0, 0, 0};
}

ContentDirectory::ContentDirectory(OCResource::Ptr resource) :
    m_searchCaps{""},
    m_sortCaps{""},
    m_id{0},
    m_resetToken{""},
    m_featureList{""},
    m_resource{resource}
{
    m_browseResult = {"", 0, 0, 0};
    m_searchResult = {"", 0, 0, 0};
}

ContentDirectory::~ContentDirectory()
{

}

ContentDirectory::ContentDirectory(const ContentDirectory &other)
{
    *this = other;
}
ContentDirectory &ContentDirectory::operator=(const ContentDirectory &other)
{
    // check for self-assignment
    if (&other == this)
    {
        return *this;
    }
    m_searchCaps = other.m_searchCaps;
    m_sortCaps = other.m_sortCaps;
    m_id = other.m_id;
    m_resetToken = other.m_resetToken;
    m_featureList = other.m_featureList;
    m_browseResult = other.m_browseResult;
    m_searchResult = other.m_searchResult;
    m_resource = other.m_resource;
    // do not copy the mutex or condition_variable
    return *this;
}

string ContentDirectory::getSearchCapabilities()
{
    unique_lock<mutex> contentDirectoryLock(m_mutex);
    m_getSearchCapabilitiesCB = bind(&ContentDirectory::onGetSearchCapabilites, this, placeholders::_1,
                                     placeholders::_2, placeholders::_3);
    m_resource->get(QueryParamsMap(), m_getSearchCapabilitiesCB);
    if (m_cv.wait_for(contentDirectoryLock,
                      chrono::seconds(MAX_WAIT_TIME_FOR_BLOCKING_CALL)) == cv_status::timeout)
    {
        cerr << "Remote device failed to respond to the request." << endl;
    }
    return m_searchCaps;
}

void ContentDirectory::onGetSearchCapabilites(const HeaderOptions &headerOptions,
        const OCRepresentation &rep, const int eCode)
{
    (void) headerOptions;
    if (eCode == OC_STACK_OK)
    {
        string uri;
        rep.getValue("uri", uri);
        if (uri == m_resource->uri())
        {
            OCRepresentation protocolInfoRep;
            rep.getValue("searchCaps", m_searchCaps);
        }
    }
    m_cv.notify_one();
}

string ContentDirectory::getSortCapabilites()
{
    unique_lock<mutex> contentDirectoryLock(m_mutex);
    m_getSortCapabilitesCB = bind(&ContentDirectory::onGetSortCapabilites, this, placeholders::_1,
                                  placeholders::_2, placeholders::_3);
    m_resource->get(QueryParamsMap(), m_getSortCapabilitesCB);
    if (m_cv.wait_for(contentDirectoryLock,
                      chrono::seconds(MAX_WAIT_TIME_FOR_BLOCKING_CALL)) == cv_status::timeout)
    {
        cerr << "Remote device failed to respond to the request." << endl;
    }
    return m_sortCaps;
}

void ContentDirectory::onGetSortCapabilites(const HeaderOptions &headerOptions,
        const OCRepresentation &rep, const int eCode)
{
    (void) headerOptions;
    if (eCode == OC_STACK_OK)
    {
        string uri;
        rep.getValue("uri", uri);
        if (uri == m_resource->uri())
        {
            OCRepresentation protocolInfoRep;
            rep.getValue("sortCaps", m_sortCaps);
        }
    }
    m_cv.notify_one();
}

int ContentDirectory::getSystemUpdateId()
{
    unique_lock<mutex> contentDirectoryLock(m_mutex);
    m_getSystemUpdateIdCB = bind(&ContentDirectory::onGetSystemUpdateId, this, placeholders::_1,
                                 placeholders::_2, placeholders::_3);
    m_resource->get(QueryParamsMap(), m_getSystemUpdateIdCB);
    if (m_cv.wait_for(contentDirectoryLock,
                      chrono::seconds(MAX_WAIT_TIME_FOR_BLOCKING_CALL)) == cv_status::timeout)
    {
        cerr << "Remote device failed to respond to the request." << endl;
    }
    return m_id;
}

void ContentDirectory::onGetSystemUpdateId(const HeaderOptions &headerOptions,
        const OCRepresentation &rep, const int eCode)
{
    (void) headerOptions;
    if (eCode == OC_STACK_OK)
    {
        string uri;
        rep.getValue("uri", uri);
        if (uri == m_resource->uri())
        {
            OCRepresentation protocolInfoRep;
            rep.getValue("systemUpdateId", m_id);
        }
    }
    m_cv.notify_one();
}

string ContentDirectory::getServiceResetToken()
{
    unique_lock<mutex> contentDirectoryLock(m_mutex);
    m_getServiceResetTokenCB = bind(&ContentDirectory::onGetServiceResetToken, this, placeholders::_1,
                                    placeholders::_2, placeholders::_3);
    m_resource->get(QueryParamsMap(), m_getServiceResetTokenCB);
    if (m_cv.wait_for(contentDirectoryLock,
                      chrono::seconds(MAX_WAIT_TIME_FOR_BLOCKING_CALL)) == cv_status::timeout)
    {
        cerr << "Remote device failed to respond to the request." << endl;
    }
    return m_resetToken;
}

void ContentDirectory::onGetServiceResetToken(const HeaderOptions &headerOptions,
        const OCRepresentation &rep, const int eCode)
{
    (void) headerOptions;
    if (eCode == OC_STACK_OK)
    {
        string uri;
        rep.getValue("uri", uri);
        if (uri == m_resource->uri())
        {
            OCRepresentation protocolInfoRep;
            rep.getValue("resetToken", m_resetToken);
        }
    }
    m_cv.notify_one();
}

string ContentDirectory::getFeatureList()
{
    unique_lock<mutex> contentDirectoryLock(m_mutex);
    m_getFeatureListCB = bind(&ContentDirectory::onGetFeatureList, this, placeholders::_1,
                              placeholders::_2, placeholders::_3);
    m_resource->get(QueryParamsMap(), m_getFeatureListCB);
    if (m_cv.wait_for(contentDirectoryLock,
                      chrono::seconds(MAX_WAIT_TIME_FOR_BLOCKING_CALL)) == cv_status::timeout)
    {
        cerr << "Remote device failed to respond to the request." << endl;
    }
    return m_featureList;
}

void ContentDirectory::onGetFeatureList(const HeaderOptions &headerOptions,
                                        const OCRepresentation &rep, const int eCode)
{
    (void) headerOptions;
    if (eCode == OC_STACK_OK)
    {
        string uri;
        rep.getValue("uri", uri);
        if (uri == m_resource->uri())
        {
            OCRepresentation protocolInfoRep;
            rep.getValue("featureList", m_featureList);
        }
    }
    m_cv.notify_one();
}

ContentDirectory::SearchResult ContentDirectory::browse(string objectId, string browseFlag,
        string filter, int startingIndex, int requestedCount, string sortCriteria)
{
    unique_lock<mutex> contentDirectoryLock(m_mutex);
    m_browseResult = {"", 0, 0, 0};
    m_browseCB = bind(&ContentDirectory::onBrowse, this, placeholders::_1,
                      placeholders::_2, placeholders::_3);
    QueryParamsMap params =
    {
        {"oid", objectId},
        {"bf", browseFlag},
        {"f", filter},
        {"si", to_string(startingIndex)},
        {"rc", to_string(requestedCount)},
        {"soc", sortCriteria}
    };
    for (auto p: params) {
        cout << p.first << " : " << p.second << endl;
    }
    m_resource->get(params, m_browseCB);
    if (m_cv.wait_for(contentDirectoryLock,
                      chrono::seconds(MAX_WAIT_TIME_FOR_BLOCKING_CALL)) == cv_status::timeout)
    {
        cerr << "Remote device failed to respond to the request." << endl;
    }
    return m_browseResult;
}

void ContentDirectory::onBrowse(const HeaderOptions &headerOptions, const OCRepresentation &rep,
                                const int eCode)
{
    (void) headerOptions;
    if (eCode == OC_STACK_OK)
    {
        string uri;
        rep.getValue("uri", uri);
        if (uri == m_resource->uri())
        {
            OCRepresentation protocolInfoRep;
            if (rep.getValue("browseResult", protocolInfoRep))
            {
                protocolInfoRep.getValue("result", m_browseResult.result);
                protocolInfoRep.getValue("numberReturned", m_browseResult.numberReturned);
                protocolInfoRep.getValue("totalMatches", m_browseResult.totalMatches);
                protocolInfoRep.getValue("updateId", m_browseResult.updateId);
            }
        }
    }
    m_cv.notify_one();
}

ContentDirectory::SearchResult ContentDirectory::search(string containerId, string searchCriteria,
        string filter, int startingIndex, int requestedCount, string sortCriteria)
{
    cout << "search containerId = " << containerId << " searchCriteria = " << searchCriteria << " filter = " << filter
            << " startIndex = " << startingIndex <<  " requestedCount = " << requestedCount << " sortCriteria = " << sortCriteria << endl;
    unique_lock<mutex> contentDirectoryLock(m_mutex);
    m_searchCB = bind(&ContentDirectory::onSearch, this, placeholders::_1,
                      placeholders::_2, placeholders::_3);
    QueryParamsMap params =
    {
        {"cid", containerId},
        {"sec", searchCriteria},
        {"f", filter},
        {"si", to_string(startingIndex)},
        {"rc", to_string(requestedCount)},
        {"soc", sortCriteria}
    };
    for (auto p: params) {
        cout << p.first << " : " << p.second << endl;
    }

    m_resource->get(params, m_searchCB);
    if (m_cv.wait_for(contentDirectoryLock,
                      chrono::seconds(MAX_WAIT_TIME_FOR_BLOCKING_CALL)) == cv_status::timeout)
    {
        cerr << "Remote device failed to respond to the request." << endl;
    }
    return m_searchResult;
}

void ContentDirectory::onSearch(const HeaderOptions &headerOptions, const OCRepresentation &rep,
                                const int eCode)
{
    (void) headerOptions;
    if (eCode == OC_STACK_OK)
    {
        string uri;
        rep.getValue("uri", uri);
        if (uri == m_resource->uri())
        {
            OCRepresentation protocolInfoRep;
            if (rep.getValue("searchResult", protocolInfoRep))
            {
                protocolInfoRep.getValue("result", m_searchResult.result);
                protocolInfoRep.getValue("numberReturned", m_searchResult.numberReturned);
                protocolInfoRep.getValue("totalMatches", m_searchResult.totalMatches);
                protocolInfoRep.getValue("updateId", m_searchResult.updateId);
            }
        }
    }
    m_cv.notify_one();
}

bool ContentDirectory::operator<(const ContentDirectory &other) const
{
    return ((*m_resource) < (*(other.m_resource)));
}

File: android/examples/upnpavclient/src/main/java/org/iotivity/base/examples/ContentDirectory.java
/*
 * //******************************************************************
 * //
 * // Copyright 2016 Intel Corporation All Rights Reserved.
 * //
 * //-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 * //
 * // Licensed under the Apache License, Version 2.0 (the "License");
 * // you may not use this file except in compliance with the License.
 * // You may obtain a copy of the License at
 * //
 * //      http://www.apache.org/licenses/LICENSE-2.0
 * //
 * // Unless required by applicable law or agreed to in writing, software
 * // distributed under the License is distributed on an "AS IS" BASIS,
 * // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * // See the License for the specific language governing permissions and
 * // limitations under the License.
 * //
 * //-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 */

package org.iotivity.base.examples;

import android.os.Parcel;
import android.os.Parcelable;

import org.iotivity.base.OcException;
import org.iotivity.base.OcRepresentation;

/**
 * ContentDirectory
 * <p/>
 * This class is used by UpnpAvClientActivity to create an object representation of a remote contentDirectory service
 * and update the values depending on the server response
 */
public class ContentDirectory extends Service implements Parcelable {

    public static final String UPNP_OIC_TYPE_CONTENT_DIRECTORY = "oic.r.content.directory";
    public static final String UPNP_OIC_URI_PREFIX_CONTENT_DIRECTORY = "/upnp/oic/content-directory/";

    public static final String SYSTEM_UPDATE_ID_KEY = "systemUpdateId";
    public static final String SEARCH_CAPS_KEY = "searchCaps";
    public static final String SORT_CAPS_KEY = "sortCaps";
    public static final String BROWSE_RESULT_KEY = "browseResult";
    public static final String SEARCH_RESULT_KEY = "searchResult";
    public static final String RESULT_KEY = "result";

    public static final String DEFAULT_SEARCH_RESULT = "";
    public static final int DEFAULT_SYSTEM_UPDATE_ID = 0;
    public static final String DEFAULT_SEARCH_CAPABILITIES = "";
    public static final String DEFAULT_SORT_CAPABILITIES = "";
    public static final String DEFAULT_BROWSE_RESULT = "";

    private int mSystemUpdateId = DEFAULT_SYSTEM_UPDATE_ID;
    private String mSearchCaps = DEFAULT_SEARCH_CAPABILITIES;
    private String mSortCaps = DEFAULT_SORT_CAPABILITIES;
    private String mBrowseResult = DEFAULT_BROWSE_RESULT;
    private String mSearchResult = DEFAULT_SEARCH_RESULT;
    private boolean mIsInitialized;

    public ContentDirectory() {
        super();
    }

    public void setOcRepresentation(OcRepresentation rep) throws OcException {
        super.setOcRepresentation(rep);
        if (rep.hasAttribute(SYSTEM_UPDATE_ID_KEY)) {
            Object obj = rep.getValue(SYSTEM_UPDATE_ID_KEY);
            if (obj instanceof Integer) {
                mSystemUpdateId = (Integer) obj;
                mIsInitialized = true;
            }
        }
        if (rep.hasAttribute(SEARCH_CAPS_KEY)) {
            Object obj = rep.getValue(SEARCH_CAPS_KEY);
            if (obj instanceof String) {
                mSearchCaps = (String) obj;
            }
        }
        if (rep.hasAttribute(SORT_CAPS_KEY)) {
            Object obj = rep.getValue(SORT_CAPS_KEY);
            if (obj instanceof String) {
                mSortCaps = (String) obj;
            }
        }
        if (rep.hasAttribute(BROWSE_RESULT_KEY)) {
            Object obj = rep.getValue(BROWSE_RESULT_KEY);
            if (obj instanceof String) {
                mBrowseResult = (String) obj;
            }
            if (obj instanceof OcRepresentation) {
                OcRepresentation embeddedOcRep = (OcRepresentation) obj;
                if (embeddedOcRep.hasAttribute(RESULT_KEY)) {
                    mBrowseResult = embeddedOcRep.getValue(RESULT_KEY);
                }
            }
        }
        if (rep.hasAttribute(SEARCH_RESULT_KEY)) {
            Object obj = rep.getValue(SEARCH_RESULT_KEY);
            if (obj instanceof String) {
                mSearchResult = (String) obj;
            }
            if (obj instanceof OcRepresentation) {
                OcRepresentation embeddedOcRep = (OcRepresentation) obj;
                if (embeddedOcRep.hasAttribute(RESULT_KEY)) {
                    mSearchResult = embeddedOcRep.getValue(RESULT_KEY);
                }
            }
        }
    }

    public OcRepresentation getOcRepresentation() throws OcException {
        OcRepresentation rep = super.getOcRepresentation();
        return rep;
    }

    public String getBrowseResult() {
        return mBrowseResult;
    }

    public String getSearchResult() {
        return mSearchResult;
    }

    public String getSearchCapabilities() {
        return mSearchCaps;
    }

    public String getSortCapabilities() {
        return mSortCaps;
    }

    public int getSystemUpdateId() {
        return mSystemUpdateId;
    }

    public boolean isInitialized() {
        return mIsInitialized;
    }

    @Override
    public String toString() {
        return "[" + super.toString() +
                ", " + "initialized: " + mIsInitialized +
                ", " + SYSTEM_UPDATE_ID_KEY + ": " + mSystemUpdateId +
                ", " + SEARCH_CAPS_KEY + ": " + mSearchCaps +
                ", " + SORT_CAPS_KEY + ": " + mSortCaps +
                ", " + BROWSE_RESULT_KEY + ": " + mBrowseResult +
                ", " + SEARCH_RESULT_KEY + ": " + mSearchResult + "]";
    }

    // Parcelable implementation
    public ContentDirectory(Parcel in) {
        readFromParcel(in);
    }

    public static final Parcelable.Creator CREATOR = new Parcelable.Creator() {
        public ContentDirectory createFromParcel(Parcel in) {
            return new ContentDirectory(in);
        }

        public ContentDirectory[] newArray(int size) {
            return new ContentDirectory[size];
        }
    };

    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(getName());
        dest.writeString(getUri());
        dest.writeInt(mSystemUpdateId);
        dest.writeString(mSearchCaps);
        dest.writeString(mSortCaps);
        dest.writeByte((byte) (mIsInitialized ? 1 : 0));
    }

    private void readFromParcel(Parcel in) {
        setName(in.readString());
        setUri(in.readString());
        mSystemUpdateId = in.readInt();
        mSearchCaps = in.readString();
        mSortCaps = in.readString();
        mIsInitialized = in.readByte() != 0;
    }

    public int describeContents() {
        return 0;
    }
}
